### ISSUE TITLE
Outputs Nan of inf with auto mixed precision

### ISSUE THREAD

**Issue by SXKai** [tensorflow/tensorflow#38495]
I use a three-tier full connection network to train a DQN (reinforcement learning) agent. The code like this

```
def observation_to_action(states):
    # define policy neural network
    W1 = tf.get_variable("W1", [state_dim, 300], initializer=tf.random_normal_initializer())
    b1 = tf.get_variable("b1", [300], initializer=tf.constant_initializer(0))
    h1 = tf.nn.relu(tf.matmul(states, W1) + b1)

    W2 = tf.get_variable("W2", [300, 300], initializer=tf.random_normal_initializer())
    b2 = tf.get_variable("b2", [300], initializer=tf.constant_initializer(0))
    h2 = tf.nn.relu(tf.matmul(h1, W2) + b2)

    W3 = tf.get_variable("W3", [300, 300], initializer=tf.random_normal_initializer())
    b3 = tf.get_variable("b3", [300], initializer=tf.constant_initializer(0))
    h3 = tf.nn.relu(tf.matmul(h2, W3) + b3)

    Wn = tf.get_variable("Wn", [300, num_actions], initializer=tf.random_normal_initializer())
    bn = tf.get_variable("bn", [num_actions], initializer=tf.constant_initializer(0))
    q = tf.matmul(h3, Wn) + bn
    return q
```

Then automatic mixed precision can be enabled by this 

```
opt = tf.train.AdamOptimizer()
opt = tf.train.experimental.enable_mixed_precision_graph_rewrite(opt)
train_op = opt.minimize(loss)
```

The log indicates that the mixed precision has been used

```
2020-04-13 09:56:57.471656: I tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1816] Running auto_mixed_precision graph optimizer
2020-04-13 09:56:57.472372: I tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1772] Converted 16/44 nodes to float16 precision using 1 cast(s) to float16 (excluding Const and Variable casts)
```

But the output of the network, q, is Nan or inf. Is this normal? I want to train a DQN agent, if the output Q-value is nan or inf, it can't choose a best action to act. As a result, it can't be trained. I think the output may overflow the range of FP16( $ 2^{-24} $ ~65504). How can I train a reinforcement learning agent with auto mixed precision. Besides this, I also try to use mix precision manually by following codes

```
def observation_to_action(states):
    # define policy neural network
    W1 = tf.get_variable("W1", [state_dim, 256], initializer=tf.random_normal_initializer())
    b1 = tf.get_variable("b1", [256], initializer=tf.constant_initializer(0))
    h1 = tf.nn.relu(tf.cast(tf.matmul(tf.cast(states, tf.float16), tf.cast(W1, tf.float16)), tf.float32) + tf.cast(b1, tf.float32))

    W2 = tf.get_variable("W2", [256, 256], initializer=tf.random_normal_initializer())
    b2 = tf.get_variable("b2", [256], initializer=tf.constant_initializer(0))
    h2 = tf.nn.relu(tf.cast(tf.matmul(tf.cast(h1, tf.float16), tf.cast(W2, tf.float16)), tf.float32) + tf.cast(b2, tf.float32))

    W3 = tf.get_variable("W3", [256, 256], initializer=tf.random_normal_initializer())
    b3 = tf.get_variable("b3", [256], initializer=tf.constant_initializer(0))
    h3 = tf.nn.relu(tf.cast(tf.matmul(tf.cast(h2, tf.float16), tf.cast(W3, tf.float16)), tf.float32) + tf.cast(b3, tf.float32))


    Wn = tf.get_variable("Wn", [256, num_actions], initializer=tf.random_normal_initializer())
    bn = tf.get_variable("bn", [num_actions], initializer=tf.constant_initializer(0))
    q = tf.cast(tf.matmul(tf.cast(h3, tf.float16), tf.cast(Wn, tf.float16)), tf.float32) + tf.cast(bn, tf.float32)

    return q
```

But it's outputs are Nan or inf too. Thank you very much.

Ubuntu 18.04.3 LTS x86_64

Tesla V100    Driver Version: 435.21       CUDA Version: 10.0

**Comment by gadagashwini** [comment]
@SXKai, Can you share the complete standalone code to reproduce the issue and also provide the Tensorflow version. Thanks

**Comment by SXKai** [comment]
> @SXKai, Can you share the complete standalone code to reproduce the issue and also provide the Tensorflow version. Thanks

Thanks. The tensorflow version I used is v1.15. I used the code from [https://github.com/yukezhu/tensorflow-reinforce](url). And I replace run_dqn_cartpole.py like following
```
from __future__ import print_function
from collections import deque

from rl.neural_q_learner import NeuralQLearner
import tensorflow as tf
import numpy as np
import gym

env_name = 'CartPole-v0'
env = gym.make(env_name)

sess = tf.Session()
optimizer = tf.train.RMSPropOptimizer(learning_rate=0.0001, decay=0.9)
optimizer = tf.train.experimental.enable_mixed_precision_graph_rewrite(optimizer)
writer = tf.summary.FileWriter("/tmp/{}-experiment-1".format(env_name))

state_dim   = env.observation_space.shape[0]
num_actions = env.action_space.n

def observation_to_action(states):
  # define policy neural network
    W1 = tf.get_variable("W1", [state_dim, 300], initializer=tf.random_normal_initializer())
    b1 = tf.get_variable("b1", [300], initializer=tf.constant_initializer(0))
    h1 = tf.nn.relu(tf.matmul(states, W1) + b1)

    W2 = tf.get_variable("W2", [300, 300], initializer=tf.random_normal_initializer())
    b2 = tf.get_variable("b2", [300], initializer=tf.constant_initializer(0))
    h2 = tf.nn.relu(tf.matmul(h1, W2) + b2)

    W3 = tf.get_variable("W3", [300, 300], initializer=tf.random_normal_initializer())
    b3 = tf.get_variable("b3", [300], initializer=tf.constant_initializer(0))
    h3 = tf.nn.relu(tf.matmul(h2, W3) + b3)

    Wn = tf.get_variable("Wn", [300, num_actions], initializer=tf.random_normal_initializer())
    bn = tf.get_variable("bn", [num_actions], initializer=tf.constant_initializer(0))
    q = tf.matmul(h3, Wn) + bn
  return q

q_learner = NeuralQLearner(sess,
                           optimizer,
                           observation_to_action,
                           state_dim,
                           num_actions,
                           summary_writer=writer)

MAX_EPISODES = 10000
MAX_STEPS    = 200

episode_history = deque(maxlen=100)
for i_episode in range(MAX_EPISODES):

  # initialize
  state = env.reset()
  total_rewards = 0

  for t in range(MAX_STEPS):
    env.render()
    action = q_learner.eGreedyAction(state[np.newaxis,:])
    next_state, reward, done, _ = env.step(action)

    total_rewards += reward
    # reward = -10 if done else 0.1 # normalize reward
    q_learner.storeExperience(state, action, reward, next_state, done)

    q_learner.updateModel()
    state = next_state

    if done: break

  episode_history.append(total_rewards)
  mean_rewards = np.mean(episode_history)

  print("Episode {}".format(i_episode))
  print("Finished after {} timesteps".format(t+1))
  print("Reward for this episode: {}".format(total_rewards))
  print("Average reward for last 100 episodes: {:.2f}".format(mean_rewards))
  if mean_rewards >= 195.0:
    print("Environment {} solved after {} episodes".format(env_name, i_episode+1))
    break
```
It can be ran by `python run_dqn_cartpole.py`.

This setting will result in Nan or inf. But I changed the initializer method of network weights ang bias as follows
```
def observation_to_action(states):
    # define policy neural network
    stdv = 1. / math.sqrt(256.)
    W1 = tf.get_variable("W1", [state_dim, 256], initializer=tf.random_uniform_initializer(minval=-stdv, maxval=stdv))
    b1 = tf.get_variable("b1", [256], initializer=tf.random_uniform_initializer(minval=-stdv, maxval=stdv))
    h1 = tf.nn.relu(tf.matmul(states, W1) + b1)
    #h1 = tf.nn.relu(tf.cast(tf.matmul(tf.cast(states, tf.float16), tf.cast(W1, tf.float16)), tf.float32) + tf.cast(b1, tf.float32))

    W2 = tf.get_variable("W2", [256, 256], initializer=tf.random_uniform_initializer(minval=-stdv, maxval=stdv))
    b2 = tf.get_variable("b2", [256], initializer=tf.random_uniform_initializer(minval=-stdv, maxval=stdv))
    h2 = tf.nn.relu(tf.matmul(h1, W2) + b2)
    #h2 = tf.nn.relu(tf.cast(tf.matmul(tf.cast(h1, tf.float16), tf.cast(W2, tf.float16)), tf.float32) + tf.cast(b2, tf.float32))

    W3 = tf.get_variable("W3", [256, 256], initializer=tf.random_uniform_initializer(minval=-stdv, maxval=stdv))
    b3 = tf.get_variable("b3", [256], initializer=tf.random_uniform_initializer(minval=-stdv, maxval=stdv))
    h3 = tf.nn.relu(tf.matmul(h2, W3) + b3)
    #h3 = tf.nn.relu(tf.cast(tf.matmul(tf.cast(h2, tf.float16), tf.cast(W3, tf.float16)), tf.float32) + tf.cast(b3, tf.float32))

    W4 = tf.get_variable("W4", [256, 256], initializer=tf.random_uniform_initializer(minval=-stdv, maxval=stdv))
    b4 = tf.get_variable("b4", [256], initializer=tf.random_uniform_initializer(minval=-stdv, maxval=stdv))
    h4 = tf.nn.relu(tf.matmul(h3, W4) + b4)

    W5 = tf.get_variable("W5", [256, 256], initializer=tf.random_uniform_initializer(minval=-stdv, maxval=stdv))
    b5 = tf.get_variable("b5", [256], initializer=tf.random_uniform_initializer(minval=-stdv, maxval=stdv))
    h5 = tf.nn.relu(tf.matmul(h4, W5) + b5)

    Wn = tf.get_variable("Wn", [256, num_actions], initializer=tf.random_uniform_initializer(minval=-stdv, maxval=stdv))
    bn = tf.get_variable("bn", [num_actions], initializer=tf.random_uniform_initializer(minval=-stdv, maxval=stdv))
    q = tf.matmul(h5, Wn) + bn
    #q2 = tf.cast(tf.matmul(tf.cast(h3, tf.float16), tf.cast(Wn, tf.float16)), tf.float32) + tf.cast(bn, tf.float32)
    return q
```
This can work and will not output Nan of inf. I think it maybe the weight is too big to overflow.

**Comment by gadagashwini** [comment]
@SXKai, I tried to reproduce the issue but received different error, Please take a look at [gist](https://colab.sandbox.google.com/gist/gadagashwini/ebd8b5922adbfa8092ae0a0d570f13e6/untitled506.ipynb) and provide more information to reproduce the issue. Thanks

**Comment by Arktius** [comment]
Check out the loss scaler [TF](https://www.tensorflow.org/api_docs/python/tf/keras/mixed_precision/experimental/LossScaleOptimizer) and the [mixed precision guide](https://www.tensorflow.org/guide/keras/mixed_precision#loss_scaling) .

**Comment by Saduf2019** [comment]
@SXKai,
please update as per above comment
